src/Kestrel.Transport.Sockets/Internal/SocketReceiver.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Net.Sockets;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    public class SocketReceiver
```

Comment:
This might be of interest:

https://github.com/dotnet/corefx/blob/2427e05827c6e6f1eeaa77b01a3596f9a03582ff/src/System.Net.Sockets/src/System/Net/Sockets/NetworkStream.cs#L964

Reply:
Defnitely similar. Is there anything you prefer about that implementation? Ours will always be slightly different. For example, we wrap our exceptions in IOExceptions after the awaiting the IO operation.

Reply:
I'm wondering if there's anywhere we can save allocations (like avoiding an extra object by deriving from SAEA)